editor/components/table-of-contents/panel.js

```diff
@@ -58,8 +56,20 @@ function TableOfContentsPanel( { blocks } ) {
 	);
 }
 
+const blockCount = ( blocksArray = [], name ) => reduce(
+	blocksArray,
+	( count, block ) =>
+		count +
+		( ! name || block.name === name ? 1 : 0 ) +
+		( block.innerBlocks ? blockCount( block.innerBlocks, name ) : 0 ),
```

Comment:
This logic is excessively complex to be inlined without comments or some named variable / separate composable function.